### PR TITLE
Fix stealth exceptions and blocking select() in daemonize() (#81064) - 2.15

### PR DESCRIPTION
**Pull request by vaygr**
* Fix stealth exceptions and blocking select() in daemonize()

* Ensure binary data transmission in daemonize

* Add changelog entries

(cherry picked from commits 894c339df8113cdb579cd00210f4ed394859a5b1, 91f94fb59d2ec3e69ff98649094f838c6d81601b)

##### SUMMARY

Backport of #81064, #81584

##### ISSUE TYPE

- Bugfix Pull Request

##### ADDITIONAL INFORMATION

N/A

### CODE REVIEW

File: lib/ansible/module_utils/service.py [hunk 1]
```diff
@@ -207,17 +207,19 @@ def daemonize(module, cmd):
         p = subprocess.Popen(run_cmd, shell=False, stdout=subprocess.PIPE, stderr=subprocess.PIPE, preexec_fn=lambda: os.close(pipe[1]))
         fds = [p.stdout, p.stderr]
 
-        # loop reading output till its done
+        # loop reading output till it is done
         output = {p.stdout: b(""), p.stderr: b("")}
         while fds:
             rfd, wfd, efd = select.select(fds, [], fds, 1)
-            if (rfd + wfd + efd) or p.poll():
+            if (rfd + wfd + efd) or p.poll() is None:
```
Comment (bcoca) [inline lib/ansible/module_utils/service.py:214]: the 'is None' changes the meaning as the existing condition would trigger either if we have output OR the process ended 'is not None, but returns an rc that is not 0', if changing this way i would suggest changing the `or` to an `and`.

File: lib/ansible/module_utils/service.py [hunk 2]
```diff
@@ -207,17 +207,19 @@ def daemonize(module, cmd):
...
                 for out in list(fds):
                     if out in rfd:
                         data = os.read(out.fileno(), chunk)
                         if not data:
                             fds.remove(out)
-                    output[out] += b(data)
+                        output[out] += data
+            else:
+                break
```
Comment (bcoca) [inline lib/ansible/module_utils/service.py:223]: I think you misunderstand why this loop is here, the while waits either until the descriptors are populated OR the program fails (poll() returns true aka rc > 0).

Comment (bcoca) [reply lib/ansible/module_utils/service.py:223]: looking back i think i made a mistake when merging the previous PR

Comment (vaygr) [reply lib/ansible/module_utils/service.py:223]: @bcoca this is a backport of #81064, I explained and verified these changes there.

Comment (vaygr) [reply lib/ansible/module_utils/service.py:223]: @bcoca both issues are very easy to reproduce: `service` module with `daemonize: true` simply doesn't work.

Comment (bcoca) [reply lib/ansible/module_utils/service.py:223]: reread that, but i'm rethinking at least part of it as the loop changes meaning this way, output assignment to data is still wrong imho, should be but in the else and then we can always ensure it is binary again.

File: lib/ansible/module_utils/service.py (hunk 1 above)
Comment (bcoca) [reply lib/ansible/module_utils/service.py:214]: `is not None` seems to better reflect the intent ... only try to read if either the descriptors are ready or the program terminated (this time including success with 0, not just errors with 1+)

File: lib/ansible/module_utils/service.py [hunk 3]
```diff
@@ -207,17 +207,19 @@ def daemonize(module, cmd):
...
             rfd, wfd, efd = select.select(fds, [], fds, 1)
-            if (rfd + wfd + efd) or p.poll():
+            if (rfd + wfd + efd) or p.poll() is None:
                 for out in list(fds):
                     if out in rfd:
                         data = os.read(out.fileno(), chunk)
                         if not data:
                             fds.remove(out)
-                    output[out] += b(data)
+                        output[out] += data
```
Comment (bcoca) [inline lib/ansible/module_utils/service.py]: in the end i think this is how this part should look
```
if data:
   output[out] += b(data)
else:
   fds.remove(out)
```

Comment (vaygr) [reply lib/ansible/module_utils/service.py]: @bcoca this will result in `'bytes' object has no attribute 'encode'`

Comment (bcoca) [reply lib/ansible/module_utils/service.py]: true, should use to_bytes() instead

Comment (bcoca) [reply lib/ansible/module_utils/service.py]: or just test for string vs bytestring .. mostly this is py2 compat code

File: lib/ansible/module_utils/service.py (hunk 1 above)
Comment (vaygr) [reply lib/ansible/module_utils/service.py:214]: @bcoca, @webknjaz no, the original condition is incorrect: that was tested. The logic here is to read if either the descriptors are ready or the program *has not* terminated.

File: lib/ansible/module_utils/service.py (hunk 3 above)
Comment (vaygr) [reply lib/ansible/module_utils/service.py]: @bcoca, I see. In this case it probably makes sense to open a corrective PR to replace that block as you suggested?

Comment (vaygr) [reply lib/ansible/module_utils/service.py]: @bcoca #81584 ready for your review.

**Comment by vaygr** [comment]
@sivel @bcoca this backport has been updated with changes from #81584.

**Pull request merged**